sql/core/src/main/scala/org/apache/spark/sql/execution/ExplainUtils.scala

```diff
@@ -0,0 +1,237 @@
+   *   1. Generates the explain output for the input plan excluding the subquery plans.
+   *   2. Generates the explain output for each subquery referenced in the plan.
+   */
+  def processPlan[T <: QueryPlan[T]](
+      plan: => QueryPlan[T],
+      append: String => Unit): Unit = {
+    try {
+      val subqueries = ArrayBuffer.empty[(SparkPlan, Expression, SparkPlan)]
+      var currentOperatorID = 0
+      currentOperatorID = processPlanSkippingSubqueries(plan, append, currentOperatorID)
+      getSubqueries(plan, subqueries)
+      var i = 0
+
+      // Process all the subqueries in the plan.
+      for (sub <- subqueries) {
+        if (i == 0) {
+          append("\n===== Subqueries =====\n\n")
+        }
+        i = i + 1
+        append(s"Subquery:$i Hosting operator id = " +
+          s"${getOpId(sub._1)} Hosting Expression = ${sub._2}\n")
+
+        // For each subquery expression in the parent plan, process its child plan to compute
+        // the explain output.
+        currentOperatorID = processPlanSkippingSubqueries(
+          sub._3,
+          append,
+          currentOperatorID)
+        append("\n")
+      }
+    } finally {
+      removeTags(plan)
+    }
+  }
+
+  /**
+   * Traverses the supplied input plan in a bottem-up fashion does the following :
+   *    1. produces a map : operator identifier -> operator
+   *    2. Records the operator id via setting a tag in the operator.
+   * Note :
+   *    1. Operator such as WholeStageCodegenExec and InputAdapter are skipped as they don't
+   *       appear in the explain output.
+   *    2. operator identifier starts at startOperatorID + 1
+   * @param plan Input query plan to process
+   * @startOperationID The start value of operation id. The subsequent operations will
+   *                  be assigned higher value.
+   * @operatorIDs A output parameter that contains a map of operator id and query plan. This
+   *             is used by caller to print the detail portion of the plan.
+   * @return The last generated operation id for this input plan. This is to ensure we
+   *         always assign incrementing unique id to each operator.
+   */
+  private def generateOperatorIDs(
+      plan: QueryPlan[_],
+      startOperatorID: Int,
+      operatorIDs: mutable.ArrayBuffer[(Int, QueryPlan[_])]): Int = {
+    var currentOperationID = startOperatorID
+    // Skip the subqueries as they are not printed as part of main query block.
+    if (plan.isInstanceOf[BaseSubqueryExec]) {
+      return currentOperationID
+    }
+    plan.foreachUp {
+      case p: WholeStageCodegenExec =>
+      case p: InputAdapter =>
+      case other: QueryPlan[_] =>
+        if (!other.getTagValue(QueryPlan.OP_ID_TAG).isDefined) {
+          currentOperationID += 1
+          other.setTagValue(QueryPlan.OP_ID_TAG, currentOperationID)
+          operatorIDs += ((currentOperationID, other))
+        }
+        other.innerChildren.foreach { plan =>
```

Comment:
if we want to skip subqueries, why handle inner children here?

Reply:
@cloud-fan Do you have a suggestion for me here ? My idea was to write this in a generic way so that i don't have to change the code that drives the recursion. The only thing we may change is at the top of the function which checks for `BaseSubqueryExec`. I wanted to make that check a function in the future that is called from both this function and `generateWholeStageCodegenIds`. This is basically to support for more cases where we want treat a operator same as `BaseSubqueryExec`.  By the way, i have examined the plan for other operators which has innerChildren such as `InsertIntoDataSourceCommand` and we are printing the plan ok.